src/MongoDB.Driver/Linq/Linq3Implementation/Ast/PipelineOptimizer/AstGroupPipelineOptimizer.cs

```diff
@@ -0,0 +1,371 @@
+﻿/* Copyright 2010-present MongoDB Inc.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+using System;
+using System.Collections.Generic;
+using MongoDB.Driver.Linq.Linq3Implementation.Ast.Expressions;
+using MongoDB.Driver.Linq.Linq3Implementation.Ast.Filters;
+using MongoDB.Driver.Linq.Linq3Implementation.Ast.Stages;
+using MongoDB.Driver.Linq.Linq3Implementation.Ast.Visitors;
+
+namespace MongoDB.Driver.Linq.Linq3Implementation.Ast.PipelineOptimizer
+{
+    internal class AstGroupPipelineOptimizer
+    {
+        private int _nextAccumulatorFieldNumber;
+        private List<AstAccumulatorField> _newAccumulatorFields = new();
+
+        public AstPipeline Optimize(AstPipeline pipeline)
+        {
+            for (var i = 0; i < pipeline.Stages.Count; i++)
+            {
+                var stage = pipeline.Stages[i];
+                if (stage is AstGroupStage groupStage)
+                {
+                    pipeline = OptimizeGroupStage(pipeline, i, groupStage);
+                }
+            }
+
+            return pipeline;
+        }
+
+        private void AddAccumulatorField(string name, AstAccumulatorExpression value)
+        {
+            _newAccumulatorFields.Add(AstExpression.AccumulatorField(name, value));
+        }
```

Comment:
To avoid the duplicate name problem, why not render `value` and see if any existing fields in `_newAccumulatorFields` match. If so, return the existing name. If not, add a new one. This method would then only take an `AstAccumulatorExpression` and return a field name.

I admit this is totally hacky and the proper way would be to walk and compare ASTs. But maybe it's good enough. Or maybe it won't work at all because of some obvious edge case that I haven't thought of.